From 29b2093462adde58e7b736ddcc2c220647abde3e Mon Sep 17 00:00:00 2001
From: Jiacheng Zheng <jiaczhen@codeaurora.org>
Date: Mon, 22 Oct 2018 13:54:37 +0800
Subject: [PATCH] kernel rootfs prefetch no dm verity

Put forward in time the mounting of rootfs and
prefetch all early applications(elf files) and
so libraries they depend on. So later when these
files are executed, the big overhead of waiting
data from emmc can now be saved.

Signed-off-by: Jiacheng Zheng <jiaczhen@codeaurora.org>
---
 drivers/md/dm-bufio.c         |  2 +-
 drivers/md/dm-verity-target.c |  2 +-
 drivers/md/dm.c               |  2 +-
 drivers/mmc/card/block.c      |  4 +++
 drivers/scsi/scsi.c           |  3 ++
 fs/ramfs/inode.c              |  2 +-
 init/Makefile                 |  2 ++
 init/do_mounts.c              | 51 +++++++++++++++++++---------
 init/initramfs.c              |  6 ++--
 init/prefetch_rootfs.c        | 79 +++++++++++++++++++++++++++++++++++++++++++
 init/prefetch_rootfs.h        | 16 +++++++++
 11 files changed, 147 insertions(+), 22 deletions(-)
 create mode 100644 init/prefetch_rootfs.c
 create mode 100644 init/prefetch_rootfs.h

diff --git a/drivers/md/dm-bufio.c b/drivers/md/dm-bufio.c
index 969c815..ceae30b 100644
--- a/drivers/md/dm-bufio.c
+++ b/drivers/md/dm-bufio.c
@@ -1906,7 +1906,7 @@ static void __exit dm_bufio_exit(void)
 		BUG();
 }
 
-module_init(dm_bufio_init)
+subsys_initcall(dm_bufio_init)
 module_exit(dm_bufio_exit)
 
 module_param_named(max_cache_size_bytes, dm_bufio_cache_size, ulong, S_IRUGO | S_IWUSR);
diff --git a/drivers/md/dm-verity-target.c b/drivers/md/dm-verity-target.c
index e34cf53..d164b41 100644
--- a/drivers/md/dm-verity-target.c
+++ b/drivers/md/dm-verity-target.c
@@ -1101,7 +1101,7 @@ static void __exit dm_verity_exit(void)
 	dm_unregister_target(&verity_target);
 }
 
-module_init(dm_verity_init);
+subsys_initcall(dm_verity_init);
 module_exit(dm_verity_exit);
 
 MODULE_AUTHOR("Mikulas Patocka <mpatocka@redhat.com>");
diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index f7f560f..c825d01 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -3761,7 +3761,7 @@ static const struct block_device_operations dm_blk_dops = {
 /*
  * module hooks
  */
-module_init(dm_init);
+subsys_initcall(dm_init);
 module_exit(dm_exit);
 
 module_param(major, uint, 0);
diff --git a/drivers/mmc/card/block.c b/drivers/mmc/card/block.c
index 2f18ad0..f9c7beb 100644
--- a/drivers/mmc/card/block.c
+++ b/drivers/mmc/card/block.c
@@ -89,6 +89,8 @@ MODULE_ALIAS("mmc:block");
 #define PCKD_TRGR_LOWER_BOUND		5
 #define PCKD_TRGR_PRECISION_MULTIPLIER	100
 
+void rootfs_prefetch(void);
+
 static struct mmc_cmdq_req *mmc_cmdq_prep_dcmd(
 		struct mmc_queue_req *mqrq, struct mmc_queue *mq);
 static DEFINE_MUTEX(block_mutex);
@@ -4718,6 +4720,8 @@ static int mmc_blk_probe(struct mmc_card *card)
 		pm_runtime_enable(&card->dev);
 	}
 
+	rootfs_prefetch();
+
 	return 0;
 
  out:
diff --git a/drivers/scsi/scsi.c b/drivers/scsi/scsi.c
index 53680ea..6688eea 100644
--- a/drivers/scsi/scsi.c
+++ b/drivers/scsi/scsi.c
@@ -1204,5 +1204,8 @@ static void __exit exit_scsi(void)
 	async_unregister_domain(&scsi_sd_probe_domain);
 }
 
+int rootfs_prefetch_wq(void);
+
 subsys_initcall(init_scsi);
+subsys_initcall(rootfs_prefetch_wq);
 module_exit(exit_scsi);
diff --git a/fs/ramfs/inode.c b/fs/ramfs/inode.c
index 889d558..892521b 100644
--- a/fs/ramfs/inode.c
+++ b/fs/ramfs/inode.c
@@ -263,4 +263,4 @@ int __init init_ramfs_fs(void)
 		return 0;
 	return register_filesystem(&ramfs_fs_type);
 }
-fs_initcall(init_ramfs_fs);
+subsys_initcall(init_ramfs_fs);
diff --git a/init/Makefile b/init/Makefile
index e96f022..b6b6d57 100644
--- a/init/Makefile
+++ b/init/Makefile
@@ -11,6 +11,8 @@ ifneq ($(CONFIG_ARCH_INIT_TASK),y)
 obj-y                          += init_task.o
 endif
 
+obj-y                          += prefetch_rootfs.o
+
 mounts-y			:= do_mounts.o
 mounts-$(CONFIG_BLK_DEV_RAM)	+= do_mounts_rd.o
 mounts-$(CONFIG_BLK_DEV_INITRD)	+= do_mounts_initrd.o
diff --git a/init/do_mounts.c b/init/do_mounts.c
index 333cc2e..5917863 100644
--- a/init/do_mounts.c
+++ b/init/do_mounts.c
@@ -547,6 +547,7 @@ void __init mount_root(void)
 #endif
 }
 
+extern int early_initrd;
 /*
  * Prepare the namespace - decide what/where to mount, load ramdisks, etc.
  */
@@ -572,23 +573,13 @@ void __init prepare_namespace(void)
 
 	md_run_setup();
 	dm_run_setup();
-	dm_verity_setup();
+	//dm_verity_setup();
 
-	if (saved_root_name[0]) {
-		root_device_name = saved_root_name;
-		if (!strncmp(root_device_name, "mtd", 3) ||
-		    !strncmp(root_device_name, "ubi", 3)) {
-			mount_block_root(root_device_name, root_mountflags);
+	if (!early_initrd) {
+		if (initrd_load())
 			goto out;
-		}
-		ROOT_DEV = name_to_dev_t(root_device_name);
-		if (strncmp(root_device_name, "/dev/", 5) == 0)
-			root_device_name += 5;
 	}
 
-	if (initrd_load())
-		goto out;
-
 	/* wait for any asynchronous scanning to complete */
 	if ((ROOT_DEV == 0) && root_wait) {
 		printk(KERN_INFO "Waiting for root device %s...\n",
@@ -604,11 +595,39 @@ void __init prepare_namespace(void)
 	if (is_floppy && rd_doload && rd_load_disk(0))
 		ROOT_DEV = Root_RAM0;
 
-	mount_root();
+	//mount_root();
 out:
 	devtmpfs_mount("dev");
-	sys_mount(".", "/", NULL, MS_MOVE, NULL);
-	sys_chroot(".");
+	if (early_initrd) {
+		sys_mount(".", "/", NULL, MS_MOVE, NULL);
+		sys_chroot(".");
+	}
+}
+
+void __init pre_prepare_namespace(void)
+{
+	dm_verity_setup();
+	if (saved_root_name[0]) {
+		root_device_name = saved_root_name;
+		if (!strncmp(root_device_name, "mtd", 3) ||
+		    !strncmp(root_device_name, "ubi", 3)) {
+			mount_block_root(root_device_name, root_mountflags);
+			goto out;
+		}
+		ROOT_DEV = name_to_dev_t(root_device_name);
+		if (strncmp(root_device_name, "/dev/", 5) == 0)
+			root_device_name += 5;
+	}
+	if (early_initrd) {
+		if (initrd_load())
+			goto out;
+	}
+	mount_root();
+out:
+	if (!early_initrd) {
+		sys_mount(".", "/", NULL, MS_MOVE, NULL);
+		sys_chroot(".");
+	}
 }
 
 static bool is_tmpfs;
diff --git a/init/initramfs.c b/init/initramfs.c
index 5205916..83db993 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -617,7 +617,8 @@ static int __init skip_initramfs_param(char *str)
 }
 __setup("skip_initramfs", skip_initramfs_param);
 
-static int __init populate_rootfs(void)
+int early_initrd = 0;
+int __init populate_rootfs(void)
 {
 	char *err;
 
@@ -634,6 +635,7 @@ static int __init populate_rootfs(void)
 #ifdef CONFIG_BLK_DEV_RAM
 		int fd;
 		printk(KERN_INFO "Trying to unpack rootfs image as initramfs...\n");
+		early_initrd = 1;
 		err = unpack_to_rootfs((char *)initrd_start,
 			initrd_end - initrd_start);
 		if (!err) {
@@ -675,4 +677,4 @@ static int __init populate_rootfs(void)
 	}
 	return 0;
 }
-rootfs_initcall(populate_rootfs);
+//rootfs_initcall(populate_rootfs);
diff --git a/init/prefetch_rootfs.c b/init/prefetch_rootfs.c
new file mode 100644
index 0000000..a178110
--- /dev/null
+++ b/init/prefetch_rootfs.c
@@ -0,0 +1,79 @@
+/* Copyright (c) 2018, The Linux Foundation. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <linux/kernel.h>
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/workqueue.h>
+
+#include "do_mounts.h"
+#include "prefetch_rootfs.h"
+#include <soc/qcom/boot_stats.h>
+
+int populate_rootfs(void);
+void pre_prepare_namespace(void);
+
+static struct workqueue_struct *prefetch_wq;
+
+int __init rootfs_prefetch_wq(void)
+{
+	prefetch_wq = alloc_workqueue("kprefetchd", WQ_HIGHPRI | WQ_UNBOUND, 
+			num_online_cpus());
+	if (!prefetch_wq) {
+		printk("Cannot allocate prefetch workqueue\n");
+	}
+	return 0;
+}
+
+static void rootfs_prefetch_work(struct work_struct *w)
+{
+	int fd, reads, i=0;
+	char *dummy_buf;
+
+	populate_rootfs();
+	pre_prepare_namespace();
+	dummy_buf = kmalloc(4096, GFP_ATOMIC);
+
+	if(dummy_buf) {
+		place_marker("M - DRIVER PREFETCH Start");
+		while(files[i])
+		{
+			fd = sys_open(files[i], O_RDONLY, 0);
+			if (fd >= 0) {
+				do{
+					reads = sys_read(fd, dummy_buf, 4096);
+				}while(reads>0);
+				sys_close(fd);
+			}
+			i++;
+		}
+		place_marker("M - DRIVER PREFETCH End");
+		kfree(dummy_buf);
+	}
+	
+	kfree(w);
+}
+
+void rootfs_prefetch(void)
+{
+	struct work_struct *prefetch_work;
+
+	if (prefetch_wq) {
+		prefetch_work = kzalloc(sizeof(struct work_struct), GFP_ATOMIC);
+		if(prefetch_work) {
+			INIT_WORK(prefetch_work, rootfs_prefetch_work);
+			queue_work(prefetch_wq, prefetch_work);
+		}
+	}
+}
+
diff --git a/init/prefetch_rootfs.h b/init/prefetch_rootfs.h
new file mode 100644
index 0000000..6471250
--- /dev/null
+++ b/init/prefetch_rootfs.h
@@ -0,0 +1,16 @@
+/* Copyright (c) 2018, The Linux Foundation. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+static char* files[] = {
+	"/usr/sbin/early_init",   //early_init
+	0};
+
-- 
1.9.1

